### PR TITLE
Allow multiline to be passed in routes when using wildcards.

### PR DESCRIPTION
**Pull request by ignacio-chiazzo**
Fixed a bug in action_dispatch where routes with newlines weren't detected when using wildcard segments

### Summary

Original issue posted: https://github.com/rails/rails/issues/39103

`Problem:` URL generation does not escape a newline character in a wildcard segment.

### Steps to reproduce it

```ruby
  draw do
    get "/wildcard/*wildcard_segment", to: SimpleApp.new("foo#index"), as: :wildcard
    get "/dynamic/:dynamic_segment", to: SimpleApp.new("foo#index"), as: :dynamic
  end

  assert_equal "/dynamic/a%0Anewline", url_helpers.dynamic_path(dynamic_segment: "a\nnewline") # succeeds
  assert_equal "/wildcard/a%0Anewline", url_helpers.wildcard_path(wildcard_segment: "a\nnewline") # fails with the error:
  # ActionController::UrlGenerationError (No route matches {:action=>"index", :controller=>"test", :wildcard_segment=>"a\nnewline"}, possible unmatched constraints: [:wildcard_segment]):
```

<details> <summary> Code snippet with tests passing </summary>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", github: 'ignacio-chiazzo/rails', branch: 'wildcard-routes'
end

require "rack/test"
require "action_controller/railtie"

class TestApp < Rails::Application
  config.root = __dir__
  config.hosts << "example.org"
  config.session_store :cookie_store, key: "cookie_store_key"
  secrets.secret_key_base = "secret_key_base"

  config.logger = Logger.new($stdout)
  Rails.logger  = config.logger

  routes.draw do
    get "/" => "test#index"
    get "/dynamic/:dynamic_segment" => "test#index", as: :dynamic
    get "/wildcard/*wildcard_segment" => "test#index", as: :wildcard
  end
end

class TestController < ActionController::Base
  include Rails.application.routes.url_helpers

  def index
    urls =
      dynamic_path(dynamic_segment: params[:s]) + "\n" +
      wildcard_path(wildcard_segment: params[:s]) + "\n"
    render plain: urls
  end
end

require "minitest/autorun"

class BugTest < Minitest::Test
  include Rack::Test::Methods

  def test_generates_url_with_spaces
    get "/?s=a%20space"
    assert last_response.ok?
    assert_equal <<~BODY, last_response.body
      /dynamic/a%20space
      /wildcard/a%20space
    BODY
  end

  def test_generates_url_with_newlines
    get "/?s=a%0Anewline"
    puts last_response
    assert last_response.ok?
    assert_equal <<~BODY, last_response.body
      /dynamic/a%0Anewline
      /wildcard/a%0Anewline
    BODY
  end

  private
    def app
      Rails.application
    end
end
```

</details>

Note: See the original issue for more details.

### Code issue:

TL;DR The issue is that the Regex doesn't check for `multiline`. 

If there is a route that uses a wildcard (e.g.`/wildcard/*wildcard_segment`) and the client sends a URL like this`/wildcard/a%0newline`, the [`Formatter` class](https://github.com/ignacio-chiazzo/rails/blob/b5f2b550f69a99336482739000c58e4e04e033aa/actionpack/lib/action_dispatch/journey/formatter.rb#L179) checks that the path `a\nnewline` matches against the REGEX `/\\A(?-mix:.+?)\\Z/`.

This REGEX uses the [Ruby opts:source](https://ruby-doc.org/core-2.2.0/Regexp.html#method-i-to_s) notation, but it's the same as `/\A#{/.+?/}\Z/`. Note that the regex embedded (`/.+?/`) doesn't use multiline (that's why we see `-mix`), which means that `multiline` (`m`), `extended mode` (`x`), and case sensitive (`i`) are disabled. 

To fix the issue I enabled multiline in the sub Regex when `wildcard` only. It looks like `wildcard` was disabled [on purpose](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/routing/mapper.rb#L256), I tried using git blame to see the reason behind but I couldn't find it (it's added a long time ago > 12 years)  [(latest commit I found)](https://github.com/ignacio-chiazzo/rails/commit/ca3936dbd691538d97d8dbe2bc1d64c21dc7db0e)

**Why does dynamic route work?**  dynamic paths don't have a `requirement`, so it doesn't check against a REGEX, [it just checks](https://github.com/ignacio-chiazzo/rails/blob/b5f2b550f69a99336482739000c58e4e04e033aa/actionpack/lib/action_dispatch/journey/formatter.rb#L173-L176) that the value `dynamic_segment` is not nil.

### Reviewers

- Is there any security issue I'm ignoring here?
- Is there any reason not to allow `multiline` when using `wildcard segments` that I am missing?

### CODE REVIEW

**Review by tenderlove — APPROVED** [review]
This looks good to me, but I'm somewhat worried about the security implications.  Especially if someone is passing this to something that shells out, or if they use the parameter in an HTTP header. @rafaelfranca do you have any thoughts?  Am I being too paranoid?

**Review by rafaelfranca — COMMENTED** [review]
I though about that as well, but aren't newline in parameters already expected? You can submit textarea fields that contains newlines.

The only difference I see is that for textarea fields you know you will likely receive newlines, but in the router parameters you will not expect. But since we accept already in dynamic segments I think it makes sense to accept in wildcard segments.

**Comment by ignacio-chiazzo** [comment]
The tests fail for some reason that I don't know 🤔
